dirvish_1_3_khl/dirvish-locate.8
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DIRVISH-LOCATE 8"
.TH DIRVISH-LOCATE 8 "2005-02-10" "perl v5.8.0" ""
.SH "NAME"
dirvish\-locate \- locate file versions in dirvish images
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdirvish-locate\fR \fIvault\fR[\fI:branch\fR] \fIpattern\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Locate versions of files in a dirvish vault
.PP
The index of each image specified \fIvault\fR is searched for paths
matching \fIpattern\fR. Each path found matching the \fIpattern\fR will be
reported followed by a modification time of each version of the file
and all images having a link to it.
.PP
The optional \fIbranch\fR specification will restrict searching to the
specified branch.
.PP
Images with an error status will be skipped as will any without index
files. The index file may be compressed by gzip or bzip2. See \fBtree\fR
and \fBindex\fR in \fB\f(BIdirvish.conf\fB\|(5)\fR for details.
.PP
The \fIpattern\fR is a \fBperl\fR regular expression to match the final
component of the path. Append \fB.*\fR to the end of the pattern if you
wish to match any substring of the whole path or \fB$\fR if you wish to
anchor the pattern to the end of the path. See \fB\f(BIperlre\fB\|(1)\fR for
details.
.PP
Directories are excluded from matching as they would wind up matching
every file within them anyway. Symlinks are also excluded from
matching.
.PP
If the \fIpattern\fR matches too many paths \fBdirvish-locate\fR will only
report the paths matched and not versions. As a sanity check if the
number of matches is really excessive \fBdirvish-locate\fR will limit the
number of images searched. Excessive matches is an indication of an
insufficiently specific \fIpattern\fR. Use the resulting path list to
compose a more specific one.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
To facilitate further automation and integration of \fBdirvish-locate\fR
with other tools \fBdirvish-locate provides rationalized exit codes.\fR
The exit codes are range based. While the code for a specific error may
change from one version to another it will remain within the specified
range. So don't test for specific exit codes but instead test for a
range of values. To the degree possible higher value ranges indicate
more severe errors.
.IP "\fB0\fR" 10
.IX Item "0"
success 200\-219 An error was encountered in loading a configuration
file.
.IP "\fB220\-254\fR" 10
.IX Item "220-254"
An error was detected in the configuration.
.IP "\fB255\fR" 10
.IX Item "255"
Incorrect usage.
.SH "FILES"
.IX Header "FILES"
\&\fB/etc/dirvish/master.conf\fR
.PP
alternate master configuration file.
.PP
\&\fB/etc/dirvish.conf\fR
.PP
master configuration file.
.PP
\&\fIbank/vault/image/\fR\fBsummary\fR
.PP
image creation summary.
.PP
\&\fIbank/vault/image/\fR\fBindex\fR
.PP
\&\fIbank/vault/image/\fR\fBindex.gz\fR
.PP
\&\fIbank/vault/image/\fR\fBindex.bz2\fR
.PP
dirvish index file.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 1
\& dirvish.conf(5)
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Dirvish was created by J.W. Schultz of Pegasystems Technologies.
.PP
Dirvish is now maintained by Keith Lofstrom at www.dirvish.org , with
the able assistance of many others.
.SH "BUGS"
.IX Header "BUGS"


dirvish_1_2/CHANGELOG
2004/02/25           dirvish.conf.5       12.0       Dirvish-1.2        
                     dirvish.pl           12.0       Dirvish-1.2        
                     install.sh           12.0       Dirvish-1.2        
                     loadconfig.pl        12.0       Dirvish-1.2        
                     RELEASE.html         12.0       Dirvish-1.2        
                     TODO.html            12.0       Dirvish-1.2        
                     dirvish-expire.8     12.0       Dirvish-1.2        
                     dirvish-expire.pl    12.0       Dirvish-1.2        
                     dirvish-locate.8     12.0       Dirvish-1.2        
                     dirvish-locate.pl    12.0       Dirvish-1.2        
                     dirvish-runall.8     12.0       Dirvish-1.2        
                     dirvish-runall.pl    12.0       Dirvish-1.2        
                     dirvish.8            12.0       Dirvish-1.2        
                     COPYING              12.0       Dirvish-1.2        
                     FAQ.html             12.0       Dirvish-1.2        
                     INSTALL              12.0       Dirvish-1.2        
                     RELEASE.html         11.7                         
                     FAQ.html             11.8                         
Preparing to tag 1.2.0

Note license change and prep for 1.2

Update license note.

2004/02/23           dirvish.pl           11.27                        
                     install.sh           11.3                         
Set --delete on rsync so retries will delete deleted files.
					(Thanks to Nathan Hunsperger)

Allow installs to existing install dir with missing subdirs.

2004/01/20           FAQ.html             11.7                         
Removed banner reference.

2004/01/19           FAQ.html             11.6                         
                     dirvish.pl           11.26                        
Reordered chmod operations on image.
					thanks to Nathan Hunsperger

2003/12/21           dirvish.pl           11.25                        
Better handling of vanished files on sender.

2003/12/17           TODO.html            11.16                        
file security check added to wish list

2003/12/15           dirvish.pl           11.24                        
Added recognition for vanishing (unexpected EOF) files on
client.  This happens due to a file being deleted or
truncated between file scan and send.

2003/12/11           FAQ.html             11.5                         
                                          11.4                         
Fixed html tag typo				-- Frank de Jong

2003/12/08           dirvish.pl           11.23                        
Fix permissions of rsync-error file.
					Thanks to Nathan Hunsperger

2003/12/06           INSTALL              11.4                         
Change to reflect rsync's 2.5.7 security release and plan
for post 2.5.6 CVS HEAD to be 2.6.0

2003/12/03           dirvish.conf.5       11.12                        
fixed some references to expire-rules (wrong plural) and
expire_rules (underscore)

2003/11/26           dirvish.conf.5       11.11                        
                     dirvish.pl           11.22                        
                                          11.21                        
                     TODO.html            11.15                        
Added a devices option (default yes) to allow excluding
devices from backups.

Fixed problems with expire-default and expire rules that
assume never.

2003/10/25           dirvish-expire.pl    11.12                        
Allow a vault named dirvish.
A vault named "tree" will still be skipped.

2003/09/17           dirvish.conf.5       11.10                        
fix typo

2003/09/14           dirvish.pl           11.20                        
                     loadconfig.pl        11.13                        
                     TODO.html            11.14                        
Check to be sure that pre and post scripts are specified
correctly (single value) in config file.

Added whitespace after seppuku_prefix

2003/09/01           TODO.html            11.13                        
                     dirvish-expire.pl    11.11                        
fixed dirvish-expire --vault so it works

Still need major rework.

2003/08/30           TODO.html            11.12                        
2003/08/29           dirvish.pl           11.19                        
                                          11.18                        
                     loadconfig.pl        11.12                        
                     TODO.html            11.11                        
                     INSTALL              11.3                         
                     FAQ.html             11.3                         
                     dirvish.conf.5       11.9                         
Fix seppuku prefix so it is $vault:$branch:$image

Added seppuku prefix variable so error messages will be
suitably marked with the vault:

Better reflect filesystem requirements in INSTALL and FAQ.

2003/08/03           dirvish-expire.pl    11.10                        
                     dirvish-locate.pl    11.9                         
                     dirvish-runall.pl    11.9                         
                     dirvish.pl           11.17                        
                     loadconfig.pl        11.11                        
                     COPYING              11.5                         
                                          11.4                         
                     dirvish-expire.pl    11.9                         
                     dirvish-locate.pl    11.8                         
                     dirvish-runall.pl    11.8                         
                     dirvish.pl           11.16                        
                     loadconfig.pl        11.10                        

dirvish_1_2/dirvish-locate.pl
#       $Id: dirvish-locate.pl,v 12.0 2004/02/25 02:42:14 jw Exp $  $Name: Dirvish-1_2 $

$VERSION = ('$Name: Dirvish-1_2 $' =~ /Dirvish/i)
	? ('$Name: Dirvish-1_2 $' =~ m/^.*:\s+dirvish-(.*)\s*\$$/i)[0]
	: '1.1.2 patch' . ('$Id: dirvish-locate.pl,v 12.0 2004/02/25 02:42:14 jw Exp $'
		=~ m/^.*,v(.*:\d\d)\s.*$/)[0];
$VERSION =~ s/_/./g;


#########################################################################
#                                                         		#
#	Copyright 2003 and $Date: 2004/02/25 02:42:14 $
#                         Pegasystems Technologies and J.W. Schultz 	#
#                                                         		#
#	Licensed under the Open Software License version 2.0		#
#                                                         		#
#	This program is free software; you can redistribute it		#
#	and/or modify it under the terms of the Open Software		#
#	License, version 2.0 by Lauwrence E. Rosen.			#
#                                                         		#
#	This program is distributed in the hope that it will be		#
#	useful, but WITHOUT ANY WARRANTY; without even the implied	#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR		#
#	PURPOSE.  See the Open Software License for details.		#
#                                                         		#
#########################################################################

use Time::ParseDate;
use POSIX qw(strftime);
use File::Find;
use Getopt::Long;

sub loadconfig;
sub check_expire;
sub findop;
sub imsort;
sub seppuku;

$KILLCOUNT = 1000;
$MAXCOUNT = 100;

sub usage
{
	my $message = shift(@_);

	length($message) and print STDERR $message, "\n\n";

	print STDERR <<EOUSAGE;
USAGE
	dirvish-locate vault[:branch] pattern
	
EOUSAGE

	exit 255;
}

$Options = 
{ 
	help		=> \&usage,
	version		=> sub {
			print STDERR "dirvish version $VERSION\n";
			exit(0);
		},
};

if ($CONFDIR =~ /dirvish$/ && -f "$CONFDIR.conf")
{
	loadconfig(undef, "$CONFDIR.conf", $Options);
}
elsif (-f "$CONFDIR/master.conf")
{
	loadconfig(undef, "$CONFDIR/master.conf", $Options);
}
elsif (-f "$CONFDIR/dirvish.conf")
{
	seppuku 250, <<EOERR;
ERROR: no master configuration file.
	An old $CONFDIR/dirvish.conf file found.
	Please read the dirvish release notes.
EOERR
}
else
{
	seppuku 251, "ERROR: no global configuration file";
}

GetOptions($Options, qw(
	version
	help|?
	)) or usage;

$Vault = shift;
$Vault =~ /:/ and ($Vault, $Branch) = split(/:/, $Vault);
$Pattern = shift;

$Vault && length($Pattern) or usage;

$fullpattern = $Pattern;
$fullpattern =~ /\$$/ or $fullpattern .= '[^/]*$';
($partpattern = $fullpattern) =~ s/^\^//;

for $b (@{$$Options{bank}})
{
	-d "$b/$Vault" and $bank = $b;
}
$bank or seppuku 220, "No such vault: $Vault";

opendir VAULT, "$bank/$Vault" or seppuku 221, "cannot open vault: $Vault";
@invault = readdir(VAULT);
closedir VAULT;

for $image (@invault)
{
	$image eq 'dirvish' and next;
	$imdir = "$bank/$Vault/$image";
	-f "$imdir/summary" or next;
	$conf = loadconfig('R', "$imdir/summary") or next;
	$$conf{Status} eq 'success' || $$conf{Status} =~ /^warn/
		or next;
	$$conf{'Backup-complete'} or next;
	$Branch && $$conf{branch} ne $Branch and next;

	unshift @images, {
		imdir => $imdir,
		image => $$conf{Image},
		branch => $$conf{branch},
		created	=> $$conf{'Backup-complete'},
	    }
}

for $image (sort(imsort @images))
{
	$imdir = $$image{imdir};

	$index = undef;
	-f "$imdir/index.bz2" and $index = "bzip2 -d -c $imdir/index.bz2|";
	-f "$imdir/index.gz" and $index = "gzip -d -c $imdir/index|";
	-f "$imdir/index" and $index = "<$imdir/index";
	$index or next;

	++$imagecount;

	open INDEX, $index or next;
	while (<INDEX>)
	{
		chomp;

		m($partpattern) or next;

# this parse operation is too slow.  It might be faster as a
# split with trimmed leading whitespace and remerged modtime
		$f = { image => $image };
		(
		 	$$f{inode},
			$$f{blocks},
			$$f{perms},
			$$f{links},
			$$f{owner},
			$$f{group},
			$$f{bytes},
			$$f{mtime},
			$path
		) = m<^
			\s*(\S+)		# inode
			\s+(\S+)		# block count
			\s+(\S+)		# perms
			\s+(\S+)		# link count
			\s+(\S+)		# owner
			\s+(\S+)		# group
			\s+(\S+)		# byte count
			\s+(\S+\s+\S+\s+\S+)	# date
			\s+(\S.*)		# path
		$>x;
		$$f{perms} =~ /^[dl]/ and next;
		$path =~ m($fullpattern) or next;

		exists($match{$path}) or ++$pathcount;
		push @{$match{$path}}, $f;
	}
	if ($pathcount >= $KILLCOUNT)
	{
		printf "dirvish-locate: too many paths match pattern, interupting search\n";
		last;
	}
}

printf "%d matches in %d images\n", $pathcount, $imagecount;
$pathcount >= $MAXCOUNT
	and printf "Pattern '%s' too vague, listing paths only.\n",
		$Pattern;

for $path (sort(keys(%match)))
{
	$last = undef;
	print $path;

	if ($pathcount >= $MAXCOUNT)
	{
		print "\n";
	       	next;
	}

	for $hit (@{$match{$path}})
	{
		$inode = $$hit{inode};
		$mtime = $$hit{mtime};
		$image = $$hit{image}{image};
		if ($inode ne $last)
		{
			$linesize = 5 + length($mtime) + length($image);
			printf "\n    %s %s", $mtime, $image;
		} else {
			$linesize += length($image) + 2;
			if ($linesize > 78)
			{
				$linesize = 5 + length($mtime) + length($image);
				print "\n",
					" " x (5 + length($mtime)),
					$image;
			} else {
				printf ", %s", $$hit{image}{image};
			}
		}
		$last = $inode;
	}
	print "\n\n";
}

exit 0;

sub imsort
{
	$$a{branch} cmp $$b{branch}
	|| $$b{created} cmp $$a{created};
}

